openebs/external-storage
# Instrument opentracing APIs in provisioner

1. Why is this change necessary ?
- To implement tracing in provisioner
- See issue : openebs/openebs#1131 for more details

2. How does this change address the issue ?
- It uses opentacing libraries to implement tracing by using contexts
  and spans.

3. How to verify this change ?
- First bring up jaeger setup on your cluster.See
  [here](https://github.com/jaegertracing/jaeger-kubernetes/)
- Open jaeger UI by getting loadbalancer/nodeport IP on browser.
- Run the templates given below
```
apiVersion: apps/v1beta1
kind: Deployment
metadata:
  name: maya-apiserver
  namespace: default
spec:
  replicas: 1
  template:
    metadata:
      labels:
        name: maya-apiserver
    spec:
      serviceAccountName: openebs-maya-operator
      containers:
      - name: maya-apiserver
        imagePullPolicy: Always
        image: utkarshmani1997/m-apiserver:testtracing1
        ports:
        - containerPort: 5656
        env:
        - name: OPENEBS_IO_JIVA_CONTROLLER_IMAGE
          value: "openebs/jiva:0.5.1-RC1"
        - name: OPENEBS_IO_JIVA_REPLICA_IMAGE
          value: "openebs/jiva:0.5.1-RC1"
        - name: OPENEBS_IO_VOLUME_MONITOR_IMAGE
          value: "openebs/m-exporter:0.5.0"
        - name: OPENEBS_IO_JIVA_REPLICA_COUNT
          value: "2"
      - image: jaegertracing/jaeger-agent
        name: jaeger-agent
        ports:
        - containerPort: 5775
          protocol: UDP
        - containerPort: 5778
        - containerPort: 6831
          protocol: UDP
        - containerPort: 6832
          protocol: UDP
        command:
        - "/go/bin/agent-linux"
        - "--collector.host-port=jaeger-collector:14267"
---
apiVersion: v1
kind: Service
metadata:
  name: maya-apiserver-service
spec:
  ports:
  - name: api
    port: 5656
    protocol: TCP
    targetPort: 5656
  selector:
    name: maya-apiserver
  sessionAffinity: None
---
apiVersion: apps/v1beta1
kind: Deployment
metadata:
  name: openebs-provisioner
  namespace: default
spec:
  replicas: 1
  template:
    metadata:
      labels:
        name: openebs-provisioner
    spec:
      serviceAccountName: openebs-maya-operator
      containers:
      - name: jaeger-agent
        imagePullPolicy: Always
        image: jaegertracing/jaeger-agent
        ports:
        - containerPort: 5775
          protocol: UDP
        - containerPort: 5778
        - containerPort: 6831
          protocol: UDP
        - containerPort: 6832
          protocol: UDP
        command:
        - "/go/bin/agent-linux"
        - "--collector.host-port=jaeger-collector:14267"
      - name: openebs-provisioner
        image: utkarshmani1997/openebs-provisioner:testtracing1
        env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
        - name: OPENEBS_MONITOR_URL
          value:
"http://127.0.0.1:32515/dashboard/db/openebs-volume-stats?orgId=1"
        - name: OPENEBS_MONITOR_VOLKEY
          value: "&var-OpenEBS"
        - name: MAYA_PORTAL_URL
          value: "https://mayaonline.io/"
```

4. What side effects does this change have ?
- none

5. Other details
- Feature : openebs/openebs#1131

Note :

- This is very first and complete PR for the provisioner, there will be
  more PR that yet be raised in openebs/maya and openebs/jiva for
  indepth tracing.

Signed-off-by: utkarshmani1997 <utkarshmani1997@gmail.com>
